### PR TITLE
feat(shared bridge): preparation for shared bridge migration (server)

### PR DESCRIPTION
**Pull request by ly0va**
## What ❔

This PR aims to reduce the diff between main and #298 to simplify the future review.

The changes do not affect current server behavior.
Only changes related to the zkSync server went into this PR.

What didn't go into this PR:

- various variable renamings
- integration tests
- changes that depend on the future version of era-contracts

Where applicable, stubs and dummy (for now) config variables were introduced, which can be replaced/populated later.

## Why ❔

Breaking up a PR into smaller chunks should be easier to review.

## Checklist

- [ ] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.
- [ ] Code has been formatted via `zk fmt` and `zk lint`.
- [ ] Spellcheck has been run via `zk spellcheck`.
- [ ] Linkcheck has been run via `zk linkcheck`.

### CODE REVIEW

File: core/lib/protobuf_config/src/contracts.rs
Comment (perekopskiy) [inline core/lib/protobuf_config/src/contracts.rs:136]: nit
```suggestion
                .context("bridgehub_impl_addr")?,
```

File: core/lib/zksync_core/src/eth_sender/eth_tx_aggregator.rs
Comment (perekopskiy) [inline core/lib/zksync_core/src/eth_sender/eth_tx_aggregator.rs:410]: Please add unit tests for post_shared_bridge eth sender

File: core/lib/zksync_core/src/eth_watch/event_processors/set_chain_id.rs
Comment (ly0va) [reply core/lib/zksync_core/src/eth_watch/event_processors/set_chain_id.rs:13]: I like this idea, this would simplify a lot of things. But don't we need to run genesis before deploying the contracts? I think it logs some config variables used in contracts later, like the genesis batch commitment etc

File: core/lib/zksync_core/src/eth_sender/tests.rs
Comment (slowli) [inline core/lib/zksync_core/src/eth_sender/tests.rs:114]: Nit: IMO, `L2ChainId::default()` would be more idiomatic here.

File: core/lib/zksync_core/src/eth_watch/event_processors/set_chain_id.rs
Comment (slowli) [inline core/lib/zksync_core/src/eth_watch/event_processors/set_chain_id.rs:62]: Nit: As I understand, this code will be refactored anyway, but this looks overly complicated. You can collect IDs into `ids: Vec<u16>` and then output its `Debug` representation via `{ids:?}`.

File: core/lib/zksync_core/src/state_keeper/io/mempool.rs
Comment (slowli) [inline core/lib/zksync_core/src/state_keeper/io/mempool.rs:458]: Dumb question: What's the reason for these changes? Do these logs become spammy with this PR? Or are they already spammy?

File: core/lib/zksync_core/src/state_keeper/keeper.rs
Comment (slowli) [inline core/lib/zksync_core/src/state_keeper/keeper.rs:250]: To repeat the question we've discussed previously: for ENs, `load_upgrade_tx()` [always returns](https://github.com/matter-labs/zksync-era/blob/758f4877433ebdd86cc90dd592db9a7d87598586/core/lib/zksync_core/src/sync_layer/external_io.rs#L544-L545) `None`.  So it seems that the EN will enter an infinite loop if `first_batch_in_shared_bridge` is true. Is `load_protocol_upgrade_tx()` not supposed to be called for ENs? Is this implemented right now?

Comment (ly0va) [reply core/lib/zksync_core/src/state_keeper/keeper.rs:250]: This loop will be removed when [Misha's suggestion](https://github.com/matter-labs/zksync-era/pull/1012#discussion_r1481209538) is implemented, so ENs should still work.

File: core/lib/zksync_core/src/state_keeper/io/mempool.rs
Comment (ly0va) [reply core/lib/zksync_core/src/state_keeper/io/mempool.rs:458]: They are already spammy :)

File: core/lib/zksync_core/src/eth_watch/event_processors/set_chain_id.rs
Comment (ly0va) [reply core/lib/zksync_core/src/eth_watch/event_processors/set_chain_id.rs:62]: Sure. It's done the same way in the upgrade even processor, I'll change it there as well

File: core/lib/zksync_core/src/eth_sender/eth_tx_aggregator.rs
Comment (ly0va) [reply core/lib/zksync_core/src/eth_sender/eth_tx_aggregator.rs:396]: Yes.

File: core/lib/zksync_core/src/state_keeper/keeper.rs
Comment (ly0va) [reply core/lib/zksync_core/src/state_keeper/keeper.rs:250]: Hmm, it will break an assertion a few lines below though. I'll fix it

File: core/bin/zksync_server/src/main.rs
Comment (RomanBrodetski) [inline core/bin/zksync_server/src/main.rs:43]: why is it configueable/optional? Is it just for the time being?

File: core/lib/dal/src/protocol_versions_dal.rs
Comment (RomanBrodetski) [inline core/lib/dal/src/protocol_versions_dal.rs:101]: nit: 
```suggestion
    async fn save_genesis_upgrade_tx_hash(&mut self, id: ProtocolVersionId, tx_hash: Option<H256>) {
```

File: core/lib/zksync_core/src/eth_sender/eth_tx_aggregator.rs
Comment (ly0va) [reply core/lib/zksync_core/src/eth_sender/eth_tx_aggregator.rs:410]: I can add some, but currently there's no point: the place where behavior differs is mocked away

File: core/lib/zksync_core/src/eth_watch/mod.rs
Comment (RomanBrodetski) [inline core/lib/zksync_core/src/eth_watch/mod.rs:98]: is this meant to be pub?

File: core/lib/zksync_core/src/lib.rs
Comment (RomanBrodetski) [inline core/lib/zksync_core/src/lib.rs:107]: what does this flag depend on? Will it be remived in the future, or will chains be able to spin up both ways?

File: core/lib/dal/src/protocol_versions_dal.rs
Comment (RomanBrodetski) [inline core/lib/dal/src/protocol_versions_dal.rs:132]: why do we combine this on the dal level? Dal shouldn't have any business logic. The fact that you need to update two tables is better represented on the app layer (`eth-watch` in this case). So please create the transaction there and call two dal methods from there

File: core/lib/zksync_core/src/state_keeper/keeper.rs
Comment (RomanBrodetski) [inline core/lib/zksync_core/src/state_keeper/keeper.rs:153]: much better in a separate method, agreed. I would also make it clear that it's an `Option` - e.g. by naming it as `protocol_upgrade_tx_option` or putting the type explicitely

File: core/lib/zksync_core/src/state_keeper/keeper.rs
Comment (RomanBrodetski) [inline core/lib/zksync_core/src/state_keeper/keeper.rs:272]: nit:

```suggestion
            tracing::info!("There is no protocol version change");
```

Also, can you please add more comments to every if clause? what's the difference between no changing protocol version and no ugprade tx to execute?

**Review by RomanBrodetski — COMMENTED** [review]
Overall looks good, mostly left some minor nits

**Review by RomanBrodetski — COMMENTED** [review]
Please wait for @perekopskiy 's oapproval and feel free to merge - unless @slowli has some request changes as well

File: core/bin/zksync_server/src/main.rs
Comment (ly0va) [reply core/bin/zksync_server/src/main.rs:43]: Firstly, there's no need for it while we don't have a shared bridge. Secondly, with the shared bridge setup, we already run `zk server --genesis` twice (first time only to generate commitments, before contract deployment, IIRC @kelemeno?). During the second time, we will run it with `--set-chain-id`.

File: core/lib/zksync_core/src/lib.rs
Comment (ly0va) [reply core/lib/zksync_core/src/lib.rs:107]: As per comment above, `zk server --genesis` will have to be run twice - with and without this flag.

File: core/lib/dal/src/protocol_versions_dal.rs
Comment (ly0va) [reply core/lib/dal/src/protocol_versions_dal.rs:132]: it's the same as [`save_protocol_version_with_tx`](https://github.com/matter-labs/zksync-era/blob/689d38c3aea4d63da259362f9f96c51c3a359191/core/lib/dal/src/protocol_versions_dal.rs#L75) though?

File: core/lib/zksync_core/src/eth_sender/eth_tx_aggregator.rs
Comment (perekopskiy) [reply core/lib/zksync_core/src/eth_sender/eth_tx_aggregator.rs:410]: I would expect from test to check that eth sender provides args compatible with ABI. I don't insist on adding test now if it would more convenient later, then feel free to add it later

File: core/lib/zksync_core/src/lib.rs
Comment (perekopskiy) [reply core/lib/zksync_core/src/lib.rs:190]: Hmm, maybe it wasn't fully clear in my initial suggestion but I had in mind to drop all the changes to eth watch from the PR and replace them with smth like this pseudo-code extracted into a separate function in `genesis.rs` file
```
let eth_client = QueryClient::new(eth_client_url);
let filter = Filter { address, topic, from_block: 0}; 
let logs = eth_client.get_logs(filter);
assert_eq!(logs.len(), 1); // There should be exactly one event we look for.
let upgrade_tx = parse(logs[0]);
dal. save_genesis_upgrade_tx(upgrade_tx);
```

File: core/lib/types/src/protocol_version.rs
Comment (slowli) [inline core/lib/types/src/protocol_version.rs:373]: Nit: This definition is duplicated from `impl TryFrom<Log> for ProtocolUpgrade` above; maybe, it makes sense to extract it to a function?

File: core/lib/zksync_core/src/eth_watch/event_processors/set_chain_id.rs
Comment (slowli) [inline core/lib/zksync_core/src/eth_watch/event_processors/set_chain_id.rs:47]: Off-topic: I feel that `LogParse` wraps a wrong type (should be some error, maybe `L1TxParseError`), but that has little to do with the PR 🙃

File: core/lib/zksync_core/src/eth_watch/mod.rs
Comment (slowli) [inline core/lib/zksync_core/src/eth_watch/mod.rs:168]: Nit: You can simplify this call chain to `.map_or(true, |l| processed_events < l)`.

File: core/lib/zksync_core/src/eth_watch/mod.rs
Comment (slowli) [inline core/lib/zksync_core/src/eth_watch/mod.rs:164]: Bikeshedding: I feel it would be better to stop the component via `stop_receiver` by either watching its side-effects (that is, the storage), or by having the component emit events that can be subscribed to. One way to achieve the latter is to add `watch::Sender<bool>` to `SetChainIDEventProcessor` (initially `false`; set to `true` once at least one event is received). This would mean you don't need to change `EventProcessor` signature etc.

File: core/lib/zksync_core/src/eth_watch/event_processors/set_chain_id.rs
Comment (slowli) [inline core/lib/zksync_core/src/eth_watch/event_processors/set_chain_id.rs:21]: Bikeshedding: Conventionally, acronyms in Rust are spelled with just the first letter capitalized (i.e., `SetChainIdEventProcessor`).

File: core/lib/zksync_core/src/eth_watch/mod.rs
Comment (slowli) [inline core/lib/zksync_core/src/eth_watch/mod.rs:251]: It feels strange that this method is non-responsive (i.e., doesn't accept `stop_receiver` from outside). Is this intentional? Is it assumed that it won't really need to wait for long realistically?

File: core/lib/zksync_core/src/lib.rs
Comment (slowli) [inline core/lib/zksync_core/src/lib.rs:621]: Typo: "netowrk_config" -> "network_config".

File: core/lib/zksync_core/src/lib.rs
Comment (ly0va) [reply core/lib/zksync_core/src/lib.rs:190]: It does indeed look simpler, but is it flexible enough? What if we send a tx with the event, but block is not validated fast enough and we miss it? I think the polling logic is handy here, and it's already implemented in `EthWatch`

File: core/lib/zksync_core/src/state_keeper/keeper.rs
Comment (slowli) [inline core/lib/zksync_core/src/state_keeper/keeper.rs:266]: Does it make sense to compare `first_tx_to_reexecute` with `protocol_upgrade_tx` more thoroughly, e.g., by hash?

File: core/lib/zksync_core/src/state_keeper/keeper.rs
Comment (slowli) [inline core/lib/zksync_core/src/state_keeper/keeper.rs:255]: Unless I'm missing something, it isn't guaranteed that `protocol_upgrade_tx` is actually `Some(_)` at this point. Should it be checked?

File: core/lib/zksync_core/src/lib.rs
Comment (perekopskiy) [reply core/lib/zksync_core/src/lib.rs:190]: Well, if event is not ready/present, then `zk server --genesis` should panic with message "Couldn't find event ..." and you rerun it when event is ready

File: core/lib/zksync_core/src/state_keeper/keeper.rs
Comment (ly0va) [reply core/lib/zksync_core/src/state_keeper/keeper.rs:255]: It is `None` if we're running external node :)

File: core/lib/zksync_core/src/eth_watch/mod.rs
Comment (ly0va) [reply core/lib/zksync_core/src/eth_watch/mod.rs:251]: Yes, this method is supposed to exit after a single `setChainId` event is found.

File: core/lib/zksync_core/src/eth_watch/mod.rs
Comment (ly0va) [reply core/lib/zksync_core/src/eth_watch/mod.rs:168]: won't I have to call an `.unwrap()` after that anyway?

File: core/lib/dal/src/protocol_versions_dal.rs
Comment (ly0va) [reply core/lib/dal/src/protocol_versions_dal.rs:115]: I think this should be done as a separate PR, since there are a lot of places in DAL where we don't propagate DB errors yet

File: core/lib/zksync_core/src/lib.rs
Comment (RomanBrodetski) [reply core/lib/zksync_core/src/lib.rs:190]: makes sense - agreed with @perekopskiy on this one

Comment (ly0va) [reply core/lib/zksync_core/src/lib.rs]: implemented

**Review by perekopskiy — DISMISSED** [review]

File: core/lib/dal/src/protocol_versions_dal.rs
Comment (RomanBrodetski) [reply core/lib/dal/src/protocol_versions_dal.rs:132]: I think both have to be reworked, but there is no need to do it now. I'm resolving the convo.

**Review by RomanBrodetski — DISMISSED** [review]
please wait for @slowli's approval

File: core/lib/zksync_core/src/genesis.rs
Comment (slowli) [inline core/lib/zksync_core/src/genesis.rs:450]: Nit: `logs.len()` would be output as the first arg in a failed `assert_eq` assertion anyway. IMO, it'd be more useful to output entire `logs` here, so that debugging is easier.

File: core/lib/zksync_core/src/genesis.rs
Comment (slowli) [inline core/lib/zksync_core/src/genesis.rs:446]: Nit: Since this function propagates errors, it could make sense to use `anyhow::ensure!(logs.len() == 1, ..)` instead.

File: core/lib/zksync_core/src/lib.rs
Comment (slowli) [inline core/lib/zksync_core/src/lib.rs:185]: Nit: Since this function propagates errors, `.context(_)` would be more idiomatic here.

File: core/lib/zksync_core/src/state_keeper/keeper.rs
Comment (slowli) [inline core/lib/zksync_core/src/state_keeper/keeper.rs:255]: Nit: Here and in other log messages in this method, it could make sense to add context (e.g., "this batch" -> "batch #{l1_batch_number}"; info about `protocol_upgrade_tx`) to simplify debugging.

File: core/lib/zksync_core/src/genesis.rs
Comment (slowli) [inline core/lib/zksync_core/src/genesis.rs:445]: Dumb question: If `save_set_chain_id_tx` is called prematurely, so that there's no event emitted yet, terminating the node is the best course of action, right?

**Review by slowli — APPROVED** [review]
There are also a couple of unresolved nits / questions from the previous review, but nothing major.

File: core/lib/zksync_core/src/genesis.rs
Comment (ly0va) [reply core/lib/zksync_core/src/genesis.rs:447]: Yes, and since this is only called during genesis, that's not a big deal at all

File: core/lib/zksync_core/src/state_keeper/keeper.rs
Comment (ly0va) [reply core/lib/zksync_core/src/state_keeper/keeper.rs:271]: since it has little to do with the current PR, let's move do it in a separate one

**Review by perekopskiy — APPROVED** [review]

**Pull request merged**
